packages/core/redux/waves/thunks.js

```diff
@@ -129,3 +145,110 @@ const sendSignedTransaction = (entry) => async (dispatch, getState) => {
+          ...wallet,
+          balances: {
+            ...wallet.balances,
+            ...formatBalances(BLOCKCHAIN_WAVES, balancesResult),
+          },
+        })
+
+        wavesProvider.subscribe(wallet.address)
+        dispatch({ type: WALLETS_SET, newWallet })
+      } catch (e) {
+        // eslint-disable-next-line no-console
+        console.log(e.message)
+      }
+    })
+    .catch((e) => {
+      // eslint-disable-next-line no-console
+      console.log('call balances from middleware is failed getWalletBalances', e)
+    })
+}
+
+export const enableWaves = () => async (dispatch) => {
+  dispatch(initToken())
+  dispatch(initWalletFromKeys())
+}
+
+const initToken = () => async (dispatch) => {
+
+  const dao = new WavesDAO(WAVES_WAVES_NAME, WAVES_WAVES_SYMBOL, wavesProvider, WAVES_DECIMALS, 'WAVES')
+  dao.watch()
+
+  const waves = await dao.fetchToken()
+  tokenService.registerDAO(waves, dao)
+  dispatch(TokensActions.tokenFetched(waves))
+  dispatch(initWavesAssetTokens(waves))
+}
+
+export const initWavesAssetTokens = (waves: TokenModel) => async (dispatch) => {
+  const assets = await wavesProvider.getAssets()
+  // do not wait until initialized, it is ok to lazy load all the tokens
+  return Promise.all(
+    Object.keys(assets)
+      .map((m) => new WavesDAO(m, m, wavesProvider, assets[m]['decimals'], assets[m]['id'], waves))
+      .map(async (dao) => {
+        try {
+          const token = await dao.fetchToken()
+          tokenService.registerDAO(token, dao)
+          dispatch(TokensActions.tokenFetched(token))
+        } catch (e) {
+          dispatch(TokensActions.tokensLoadingFailed())
+        }
+      }),
+  )
+}
+
+const initWalletFromKeys = () => async (dispatch, getState) => {
+  const state = getState()
+  const { network } = getCurrentNetworkSelector(state)
+  const addressCache = { ...getAddressCache(state) }
+
+  if (!addressCache[BLOCKCHAIN_WAVES]) {
+    const path = WavesUtils.getWavesDerivedPath(network[BLOCKCHAIN_WAVES])
+    const signer = getWavesSigner(state)
+
+    if (signer) {
+      const address = await signer.getAddress(path)
+      addressCache[BLOCKCHAIN_WAVES] = {
+        address,
+        path,
+      }
+
+      dispatch({
+        type: WALLETS_CACHE_ADDRESS,
+        blockchain: BLOCKCHAIN_WAVES,
+        address,
+        path,
+      })
+    }
+  }
+
+  const { address, path } = addressCache[BLOCKCHAIN_WAVES]
+  const wallet = new WalletModel({
+    address,
+    blockchain: BLOCKCHAIN_WAVES,
+    isMain: true,
+    walletDerivedPath: path,
+  })
+
+  wavesProvider.subscribe(wallet.address)
+  dispatch({ type: WALLETS_SET, wallet })
+  dispatch(updateWalletBalance(wallet))
+}
+
+export const disableWaves = () => async (dispatch, getState) => {
+  const wallets = getWalletsByBlockchain(BLOCKCHAIN_WAVES)(getState())
+
+  wallets.forEach((wallet) => {
+    wavesProvider.unsubscribe(wallet.address)
+    dispatch({ type: WALLETS_UNSET, wallet })
+  })
+}
+
```

Comment:
Much logic is duplicated among blockchain thunks. The duplication should be vanished.

Reply:
Need more details